### ISSUE TITLE
Campaign not getting saved

### ISSUE THREAD

**Issue by Allen-Paul** [securestate/king-phisher#399]
I'm currently facing an issue while I try to save a campaign, I tried reinstalling the whole app along with their dependencies and no satisfactory results were obtained.
I even tried manually going into POSTGRESQL db and removing the king-phisher table, but none of them has helped me resolve it.

What happens is like this

I create a new campaign with all details and on the final page when I press finish I get a window which says unhandled exception has occurred

--------------------------------------------------------------------------------------------------------

Error Type: builtins.KeyError
Error Details: KeyError('hasSni',)
Error UID: d60eeaad-157f-499e-b8a4-87a187b1cd02
RPC Error: N/A (Not an RPC error)
King Phisher Version: 1.14.1 (rev: 52dfd0acb6d0)
Platform Version: Linux: Ubuntu 18.04 bionic
Python Version: 3.6.8
Gtk Version: 3.22.30
Timezone: Etc/UTC

Thread Information:
===================
 => MainThread (alive=True daemon=False)
    RPCAsyncWorker (alive=True daemon=False)
    Thread-1 (alive=True daemon=True)

Stack Trace:
============
Traceback (most recent call last):
  File "/opt/king-phisher/king_phisher/client/assistants/campaign.py", line 645, in signal_assistant_apply
    if self._ssl_status['hasSni']:
KeyError: 'hasSni'

--------------------------------------------------------------------------------------------------------

# Project issue tracker-error details as follows

--------------------------------------------------------------------------------------------------------

Error Type: builtins.RuntimeError
Error Details: RuntimeError('could not find suitable application to open uri',)
Error UID: ffc8913a-bce6-4693-8bc8-49a2bbbe6d4e
RPC Error: N/A (Not an RPC error)
King Phisher Version: 1.14.1 (rev: 52dfd0acb6d0)
Platform Version: Linux: Ubuntu 18.04 bionic
Python Version: 3.6.8
Gtk Version: 3.22.30
Timezone: Etc/UTC

Thread Information:
===================
 => MainThread (alive=True daemon=False)
    RPCAsyncWorker (alive=True daemon=False)
    Thread-1 (alive=True daemon=True)

Stack Trace:
============
Traceback (most recent call last):
  File "/opt/king-phisher/king_phisher/client/dialogs/exception.py", line 177, in <lambda>
    linkbutton.connect('activate-link', lambda _: utilities.open_uri(linkbutton.get_property('uri')))
  File "/opt/king-phisher/king_phisher/utilities.py", line 331, in open_uri
    raise RuntimeError('could not find suitable application to open uri')
RuntimeError: could not find suitable application to open uri

--------------------------------------------------------------------------------------------------------

# then I press okay, I see the campaign is already saved as it asks me to create a unique campaign name.
# So I press cancel and I  go to 'open campaign window
# I select the campaign which I just created to see hostname and path is missing the websever URL tab.
# no matter whatever I do the URL and path doesn't get saved.

# I'm also including the DEBUG serverconfig.yml below

--------------------------------------------------------------------------------------------------------

INFO     connecting to local king phisher instance
DEBUG    using serializer: application/json
DEBUG    using serializer: binary/message-pack
DEBUG    thread RPCAsyncWorker running king_phisher.client.client_rpc._async_thread_routine in tid: 0x7fa4b8ffe700
DEBUG    calling RPC method: version
DEBUG    the async RPC worker has started
INFO     successfully connected to the king phisher server (version: 1.14.1 (rev: 52dfd0acb6d0) rpc api version: 6.5)
DEBUG    calling RPC method: login
DEBUG    successfully authenticated to the remote king phisher service
INFO     connecting to the server event socket
DEBUG    thread Thread-1 running websocket._app.run_forever in tid: 0x7fa4a9fff700
DEBUG    successfully connected to the server event socket
DEBUG    calling RPC method: graphql
DEBUG    calling RPC method: config/get
DEBUG    calling RPC method: graphql
DEBUG    loading GTK builder object with id: CampaignSelectionDialog.button_new_campaign
DEBUG    loading GTK builder object with id: CampaignSelectionDialog.button_select
DEBUG    loading GTK builder object with id: CampaignSelectionDialog.drawingarea_color_key
DEBUG    loading GTK builder object with id: CampaignSelectionDialog.label_campaign_info
DEBUG    loading GTK builder object with id: CampaignSelectionDialog.menubutton_filter
DEBUG    loading GTK builder object with id: CampaignSelectionDialog.treeview_campaigns
DEBUG    calling RPC method: graphql
DEBUG    loading GTK builder object with id: CompanyEditorGrid.combobox_company_industry
DEBUG    loading GTK builder object with id: CompanyEditorGrid.entry_company_industry
DEBUG    loading GTK builder object with id: CompanyEditorGrid.entry_company_name
DEBUG    loading GTK builder object with id: CompanyEditorGrid.entry_company_description
DEBUG    loading GTK builder object with id: CompanyEditorGrid.entry_company_url_main
DEBUG    loading GTK builder object with id: CompanyEditorGrid.entry_company_url_email
DEBUG    loading GTK builder object with id: CompanyEditorGrid.entry_company_url_remote_access
DEBUG    loading GTK builder object with id: CampaignAssistant.alignment_company
DEBUG    loading GTK builder object with id: CampaignAssistant.button_select_kpm_dest_folder
DEBUG    loading GTK builder object with id: CampaignAssistant.button_select_kpm_file
DEBUG    loading GTK builder object with id: CampaignAssistant.button_url_ssl_issue_certificate
DEBUG    loading GTK builder object with id: CampaignAssistant.calendar_campaign_expiration
DEBUG    loading GTK builder object with id: CampaignAssistant.checkbutton_alert_subscribe
DEBUG    loading GTK builder object with id: CampaignAssistant.checkbutton_expire_campaign
DEBUG    loading GTK builder object with id: CampaignAssistant.checkbutton_reject_after_credentials
DEBUG    loading GTK builder object with id: CampaignAssistant.combobox_campaign_type
DEBUG    loading GTK builder object with id: CampaignAssistant.combobox_company_existing
DEBUG    loading GTK builder object with id: CampaignAssistant.combobox_url_hostname
DEBUG    loading GTK builder object with id: CampaignAssistant.combobox_url_path
DEBUG    loading GTK builder object with id: CampaignAssistant.combobox_url_scheme
DEBUG    loading GTK builder object with id: CampaignAssistant.entry_campaign_description
DEBUG    loading GTK builder object with id: CampaignAssistant.entry_campaign_name
DEBUG    loading GTK builder object with id: CampaignAssistant.entry_kpm_dest_folder
DEBUG    loading GTK builder object with id: CampaignAssistant.entry_kpm_file
DEBUG    loading GTK builder object with id: CampaignAssistant.entry_test_validation_text
DEBUG    loading GTK builder object with id: CampaignAssistant.entry_validation_regex_mfa_token
DEBUG    loading GTK builder object with id: CampaignAssistant.entry_validation_regex_password
DEBUG    loading GTK builder object with id: CampaignAssistant.entry_validation_regex_username
DEBUG    loading GTK builder object with id: CampaignAssistant.frame_campaign_expiration
DEBUG    loading GTK builder object with id: CampaignAssistant.frame_company_existing
DEBUG    loading GTK builder object with id: CampaignAssistant.frame_company_new
DEBUG    loading GTK builder object with id: CampaignAssistant.image_intro_title
DEBUG    loading GTK builder object with id: CampaignAssistant.image_url_ssl_status
DEBUG    loading GTK builder object with id: CampaignAssistant.label_confirm_body
DEBUG    loading GTK builder object with id: CampaignAssistant.label_confirm_title
DEBUG    loading GTK builder object with id: CampaignAssistant.label_intro_body
DEBUG    loading GTK builder object with id: CampaignAssistant.label_intro_title
DEBUG    loading GTK builder object with id: CampaignAssistant.label_url_for_scheme
DEBUG    loading GTK builder object with id: CampaignAssistant.label_url_for_hostname
DEBUG    loading GTK builder object with id: CampaignAssistant.label_url_for_path
DEBUG    loading GTK builder object with id: CampaignAssistant.label_url_info_authors
DEBUG    loading GTK builder object with id: CampaignAssistant.label_url_info_created
DEBUG    loading GTK builder object with id: CampaignAssistant.label_url_info_description
DEBUG    loading GTK builder object with id: CampaignAssistant.label_url_info_title
DEBUG    loading GTK builder object with id: CampaignAssistant.label_url_info_for_authors
DEBUG    loading GTK builder object with id: CampaignAssistant.label_url_preview
DEBUG    loading GTK builder object with id: CampaignAssistant.label_url_ssl_for_status
DEBUG    loading GTK builder object with id: CampaignAssistant.label_url_ssl_status
DEBUG    loading GTK builder object with id: CampaignAssistant.label_validation_regex_mfa_token
DEBUG    loading GTK builder object with id: CampaignAssistant.label_validation_regex_password
DEBUG    loading GTK builder object with id: CampaignAssistant.label_validation_regex_username
DEBUG    loading GTK builder object with id: CampaignAssistant.listbox_url_info_classifiers
DEBUG    loading GTK builder object with id: CampaignAssistant.listbox_url_info_references
DEBUG    loading GTK builder object with id: CampaignAssistant.radiobutton_company_existing
DEBUG    loading GTK builder object with id: CampaignAssistant.radiobutton_company_new
DEBUG    loading GTK builder object with id: CampaignAssistant.radiobutton_company_none
DEBUG    loading GTK builder object with id: CampaignAssistant.revealer_url_ssl_settings
DEBUG    loading GTK builder object with id: CampaignAssistant.togglebutton_expiration_time
DEBUG    setting proxied widget CompanyEditorGrid via alignment_company.add
DEBUG    calling RPC method: graphql
DEBUG    loading GTK builder object with id: TimeSelector.spinbutton_hour
DEBUG    loading GTK builder object with id: TimeSelector.spinbutton_minute
DEBUG    calling RPC method: graphql
DEBUG    calling RPC method: graphql
DEBUG    calling RPC method: graphql
DEBUG    calling RPC method: config/get
DEBUG    calling RPC method: hostnames/get
DEBUG    calling RPC method: graphql
ERROR    async rpc method: graphql encountered an error
Traceback (most recent call last):
  File "/opt/king-phisher/king_phisher/client/client_rpc.py", line 242, in _async_thread_routine
    results = work_item.method(*args, **kwargs)
  File "/opt/king-phisher/king_phisher/client/client_rpc.py", line 364, in graphql
    query_vars=query_vars
king_phisher.errors.KingPhisherGraphQLQueryError
DEBUG    loading GTK builder object with id: CompanyEditorGrid.combobox_company_industry
DEBUG    loading GTK builder object with id: CompanyEditorGrid.entry_company_industry
DEBUG    loading GTK builder object with id: CompanyEditorGrid.entry_company_name
DEBUG    loading GTK builder object with id: CompanyEditorGrid.entry_company_description
DEBUG    loading GTK builder object with id: CompanyEditorGrid.entry_company_url_main
DEBUG    loading GTK builder object with id: CompanyEditorGrid.entry_company_url_email
DEBUG    loading GTK builder object with id: CompanyEditorGrid.entry_company_url_remote_access
DEBUG    loading GTK builder object with id: CampaignAssistant.alignment_company
DEBUG    loading GTK builder object with id: CampaignAssistant.button_select_kpm_dest_folder
DEBUG    loading GTK builder object with id: CampaignAssistant.button_select_kpm_file
DEBUG    loading GTK builder object with id: CampaignAssistant.button_url_ssl_issue_certificate
DEBUG    loading GTK builder object with id: CampaignAssistant.calendar_campaign_expiration
DEBUG    loading GTK builder object with id: CampaignAssistant.checkbutton_alert_subscribe
DEBUG    loading GTK builder object with id: CampaignAssistant.checkbutton_expire_campaign
DEBUG    loading GTK builder object with id: CampaignAssistant.checkbutton_reject_after_credentials
DEBUG    loading GTK builder object with id: CampaignAssistant.combobox_campaign_type
DEBUG    loading GTK builder object with id: CampaignAssistant.combobox_company_existing
DEBUG    loading GTK builder object with id: CampaignAssistant.combobox_url_hostname
DEBUG    loading GTK builder object with id: CampaignAssistant.combobox_url_path
DEBUG    loading GTK builder object with id: CampaignAssistant.combobox_url_scheme
DEBUG    loading GTK builder object with id: CampaignAssistant.entry_campaign_description
DEBUG    loading GTK builder object with id: CampaignAssistant.entry_campaign_name
DEBUG    loading GTK builder object with id: CampaignAssistant.entry_kpm_dest_folder
DEBUG    loading GTK builder object with id: CampaignAssistant.entry_kpm_file
DEBUG    loading GTK builder object with id: CampaignAssistant.entry_test_validation_text
DEBUG    loading GTK builder object with id: CampaignAssistant.entry_validation_regex_mfa_token
DEBUG    loading GTK builder object with id: CampaignAssistant.entry_validation_regex_password
DEBUG    loading GTK builder object with id: CampaignAssistant.entry_validation_regex_username
DEBUG    loading GTK builder object with id: CampaignAssistant.frame_campaign_expiration
DEBUG    loading GTK builder object with id: CampaignAssistant.frame_company_existing
DEBUG    loading GTK builder object with id: CampaignAssistant.frame_company_new
DEBUG    loading GTK builder object with id: CampaignAssistant.image_intro_title
DEBUG    loading GTK builder object with id: CampaignAssistant.image_url_ssl_status
DEBUG    loading GTK builder object with id: CampaignAssistant.label_confirm_body
DEBUG    loading GTK builder object with id: CampaignAssistant.label_confirm_title
DEBUG    loading GTK builder object with id: CampaignAssistant.label_intro_body
DEBUG    loading GTK builder object with id: CampaignAssistant.label_intro_title
DEBUG    loading GTK builder object with id: CampaignAssistant.label_url_for_scheme
DEBUG    loading GTK builder object with id: CampaignAssistant.label_url_for_hostname
DEBUG    loading GTK builder object with id: CampaignAssistant.label_url_for_path
DEBUG    loading GTK builder object with id: CampaignAssistant.label_url_info_authors
DEBUG    loading GTK builder object with id: CampaignAssistant.label_url_info_created
DEBUG    loading GTK builder object with id: CampaignAssistant.label_url_info_description
DEBUG    loading GTK builder object with id: CampaignAssistant.label_url_info_title
DEBUG    loading GTK builder object with id: CampaignAssistant.label_url_info_for_authors
DEBUG    loading GTK builder object with id: CampaignAssistant.label_url_preview
DEBUG    loading GTK builder object with id: CampaignAssistant.label_url_ssl_for_status
DEBUG    loading GTK builder object with id: CampaignAssistant.label_url_ssl_status
DEBUG    loading GTK builder object with id: CampaignAssistant.label_validation_regex_mfa_token
DEBUG    loading GTK builder object with id: CampaignAssistant.label_validation_regex_password
DEBUG    loading GTK builder object with id: CampaignAssistant.label_validation_regex_username
DEBUG    loading GTK builder object with id: CampaignAssistant.listbox_url_info_classifiers
DEBUG    loading GTK builder object with id: CampaignAssistant.listbox_url_info_references
DEBUG    loading GTK builder object with id: CampaignAssistant.radiobutton_company_existing
DEBUG    loading GTK builder object with id: CampaignAssistant.radiobutton_company_new
DEBUG    loading GTK builder object with id: CampaignAssistant.radiobutton_company_none
DEBUG    loading GTK builder object with id: CampaignAssistant.revealer_url_ssl_settings
DEBUG    loading GTK builder object with id: CampaignAssistant.togglebutton_expiration_time
DEBUG    setting proxied widget CompanyEditorGrid via alignment_company.add
DEBUG    calling RPC method: graphql
DEBUG    loading GTK builder object with id: TimeSelector.spinbutton_hour
DEBUG    loading GTK builder object with id: TimeSelector.spinbutton_minute
DEBUG    calling RPC method: graphql
DEBUG    calling RPC method: graphql
DEBUG    calling RPC method: graphql
DEBUG    calling RPC method: config/get
DEBUG    calling RPC method: hostnames/get
DEBUG    calling RPC method: graphql
ERROR    async rpc method: graphql encountered an error
Traceback (most recent call last):
  File "/opt/king-phisher/king_phisher/client/client_rpc.py", line 242, in _async_thread_routine
    results = work_item.method(*args, **kwargs)
  File "/opt/king-phisher/king_phisher/client/client_rpc.py", line 364, in graphql
    query_vars=query_vars
king_phisher.errors.KingPhisherGraphQLQueryError
DEBUG    loading GTK builder object with id: CompanyEditorGrid.combobox_company_industry
DEBUG    loading GTK builder object with id: CompanyEditorGrid.entry_company_industry
DEBUG    loading GTK builder object with id: CompanyEditorGrid.entry_company_name
DEBUG    loading GTK builder object with id: CompanyEditorGrid.entry_company_description
DEBUG    loading GTK builder object with id: CompanyEditorGrid.entry_company_url_main
DEBUG    loading GTK builder object with id: CompanyEditorGrid.entry_company_url_email
DEBUG    loading GTK builder object with id: CompanyEditorGrid.entry_company_url_remote_access
DEBUG    loading GTK builder object with id: CampaignAssistant.alignment_company
DEBUG    loading GTK builder object with id: CampaignAssistant.button_select_kpm_dest_folder
DEBUG    loading GTK builder object with id: CampaignAssistant.button_select_kpm_file
DEBUG    loading GTK builder object with id: CampaignAssistant.button_url_ssl_issue_certificate
DEBUG    loading GTK builder object with id: CampaignAssistant.calendar_campaign_expiration
DEBUG    loading GTK builder object with id: CampaignAssistant.checkbutton_alert_subscribe
DEBUG    loading GTK builder object with id: CampaignAssistant.checkbutton_expire_campaign
DEBUG    loading GTK builder object with id: CampaignAssistant.checkbutton_reject_after_credentials
DEBUG    loading GTK builder object with id: CampaignAssistant.combobox_campaign_type
DEBUG    loading GTK builder object with id: CampaignAssistant.combobox_company_existing
DEBUG    loading GTK builder object with id: CampaignAssistant.combobox_url_hostname
DEBUG    loading GTK builder object with id: CampaignAssistant.combobox_url_path
DEBUG    loading GTK builder object with id: CampaignAssistant.combobox_url_scheme
DEBUG    loading GTK builder object with id: CampaignAssistant.entry_campaign_description
DEBUG    loading GTK builder object with id: CampaignAssistant.entry_campaign_name
DEBUG    loading GTK builder object with id: CampaignAssistant.entry_kpm_dest_folder
DEBUG    loading GTK builder object with id: CampaignAssistant.entry_kpm_file
DEBUG    loading GTK builder object with id: CampaignAssistant.entry_test_validation_text
DEBUG    loading GTK builder object with id: CampaignAssistant.entry_validation_regex_mfa_token
DEBUG    loading GTK builder object with id: CampaignAssistant.entry_validation_regex_password
DEBUG    loading GTK builder object with id: CampaignAssistant.entry_validation_regex_username
DEBUG    loading GTK builder object with id: CampaignAssistant.frame_campaign_expiration
DEBUG    loading GTK builder object with id: CampaignAssistant.frame_company_existing
DEBUG    loading GTK builder object with id: CampaignAssistant.frame_company_new
DEBUG    loading GTK builder object with id: CampaignAssistant.image_intro_title
DEBUG    loading GTK builder object with id: CampaignAssistant.image_url_ssl_status
DEBUG    loading GTK builder object with id: CampaignAssistant.label_confirm_body
DEBUG    loading GTK builder object with id: CampaignAssistant.label_confirm_title
DEBUG    loading GTK builder object with id: CampaignAssistant.label_intro_body
DEBUG    loading GTK builder object with id: CampaignAssistant.label_intro_title
DEBUG    loading GTK builder object with id: CampaignAssistant.label_url_for_scheme
DEBUG    loading GTK builder object with id: CampaignAssistant.label_url_for_hostname
DEBUG    loading GTK builder object with id: CampaignAssistant.label_url_for_path
DEBUG    loading GTK builder object with id: CampaignAssistant.label_url_info_authors
DEBUG    loading GTK builder object with id: CampaignAssistant.label_url_info_created
DEBUG    loading GTK builder object with id: CampaignAssistant.label_url_info_description
DEBUG    loading GTK builder object with id: CampaignAssistant.label_url_info_title
DEBUG    loading GTK builder object with id: CampaignAssistant.label_url_info_for_authors
DEBUG    loading GTK builder object with id: CampaignAssistant.label_url_preview
DEBUG    loading GTK builder object with id: CampaignAssistant.label_url_ssl_for_status
DEBUG    loading GTK builder object with id: CampaignAssistant.label_url_ssl_status
DEBUG    loading GTK builder object with id: CampaignAssistant.label_validation_regex_mfa_token
DEBUG    loading GTK builder object with id: CampaignAssistant.label_validation_regex_password
DEBUG    loading GTK builder object with id: CampaignAssistant.label_validation_regex_username
DEBUG    loading GTK builder object with id: CampaignAssistant.listbox_url_info_classifiers
DEBUG    loading GTK builder object with id: CampaignAssistant.listbox_url_info_references
DEBUG    loading GTK builder object with id: CampaignAssistant.radiobutton_company_existing
DEBUG    loading GTK builder object with id: CampaignAssistant.radiobutton_company_new
DEBUG    loading GTK builder object with id: CampaignAssistant.radiobutton_company_none
DEBUG    loading GTK builder object with id: CampaignAssistant.revealer_url_ssl_settings
DEBUG    loading GTK builder object with id: CampaignAssistant.togglebutton_expiration_time
DEBUG    setting proxied widget CompanyEditorGrid via alignment_company.add
DEBUG    calling RPC method: graphql
DEBUG    loading GTK builder object with id: TimeSelector.spinbutton_hour
DEBUG    loading GTK builder object with id: TimeSelector.spinbutton_minute
DEBUG    calling RPC method: graphql
DEBUG    calling RPC method: graphql
DEBUG    calling RPC method: graphql
DEBUG    calling RPC method: config/get
DEBUG    calling RPC method: hostnames/get
DEBUG    calling RPC method: graphql
ERROR    async rpc method: graphql encountered an error
Traceback (most recent call last):
  File "/opt/king-phisher/king_phisher/client/client_rpc.py", line 242, in _async_thread_routine
    results = work_item.method(*args, **kwargs)
  File "/opt/king-phisher/king_phisher/client/client_rpc.py", line 364, in graphql
    query_vars=query_vars
king_phisher.errors.KingPhisherGraphQLQueryError
DEBUG    calling RPC method: graphql
DEBUG    calling RPC method: graphql
DEBUG    calling RPC method: graphql
DEBUG    calling RPC method: graphql
DEBUG    calling RPC method: graphql
DEBUG    calling RPC method: graphql
DEBUG    calling RPC method: graphql
DEBUG    calling RPC method: graphql
DEBUG    calling RPC method: graphql
DEBUG    calling RPC method: graphql
DEBUG    calling RPC method: graphql
DEBUG    calling RPC method: db/table/insert
DEBUG    calling RPC method: db/table/insert
DEBUG    calling RPC method: campaign/new
DEBUG    calling RPC method: db/table/set
DEBUG    calling RPC method: campaign/alerts/is_subscribed
DEBUG    calling RPC method: graphql
ERROR    error uid: 393a2095-570e-4b52-998b-dbe8a253cee2 an unhandled exception was thrown
Traceback (most recent call last):
  File "/opt/king-phisher/king_phisher/client/assistants/campaign.py", line 645, in signal_assistant_apply
    if self._ssl_status['hasSni']:
KeyError: 'hasSni'
DEBUG    loading GTK builder object with id: ExceptionDialog.linkbutton_github_issues
DEBUG    loading GTK builder object with id: ExceptionDialog.label_error_description
DEBUG    loading GTK builder object with id: ExceptionDialog.textview_error_details
DEBUG    idle loader routine called
DEBUG    thread Thread-2 running king_phisher.client.tabs.campaign.loader_thread_routine in tid: 0x7fa4977fe700
DEBUG    calling RPC method: graphql
DEBUG    calling RPC method: graphql
DEBUG    calling RPC method: ping

# I hope there is a resolution to this as soon as possible
# Regards.

**Comment by zeroSteiner** [comment]
The URL isn't supposed to be saved in the database. It get saved locally by the client. If you open the campaign from another host using King Phisher, that option won't be set. When you configure the campaign, the Webserver URL option is put into the Message tab.

You're then reporting two separate stack traces. The first looks like a bug while the second is complaining that it can't find an application to open the URL.

**Comment by wolfthefallen** [comment]
I was able to create the the stack trace on the master branch.
```
Stack Trace:
============
Traceback (most recent call last):
  File "/opt/king-phisher/king_phisher/client/assistants/campaign.py", line 645, in signal_assistant_apply
    if self._ssl_status['hasSni']:
KeyError: 'hasSni'
```

**Comment by zeroSteiner** [comment]
@wolfthefallen reproduction steps please.

**Comment by wolfthefallen** [comment]
* Create a fresh install of King Phisher on master.
  * Drop your king_phisher database to make sure it is empty
* start the King Phisher server. It might error out when creating the new database, start the server again and it will find the new database it created.
* start the client and log in
* create new campaign
* On the URL put in something such as `local.king-phisher.com` and path such as `\index.html`
* click through to finish and accept
* stack trace

**Comment by zeroSteiner** [comment]
Those steps didn't work for me. The key error is also complaining about a dictionary that get's populated with an async GraphQL query, so I don't know how clearing the database would affect that. My best guess is that it's a race condition in how fast you can hit apply before the data is loaded.

@wolfthefallen can you reliably reproduce it following those steps and what if you deliberately wait a couple of seconds before hitting Apply to finish editing the campaign?

**Comment by wolfthefallen** [comment]
the GraphQL query is getting a stack track. It is erroring out on Certbot path which I current do not have in my server config. 

Server Side
```
DEBUG    127.0.0.1:45116 tid: 0x7ff44f5f6700 running http request handler
ERROR    Traceback (most recent call last):
  File "/opt/king-phisher/.venv/lib/python3.6/site-packages/promise/promise.py", line 487, in _resolve_from_executor
    executor(resolve, reject)
  File "/opt/king-phisher/.venv/lib/python3.6/site-packages/promise/promise.py", line 754, in executor
    return resolve(f(*args, **kwargs))
  File "/opt/king-phisher/.venv/lib/python3.6/site-packages/graphql/execution/middleware.py", line 75, in make_it_promise
    return next(*args, **kwargs)
  File "/opt/king-phisher/king_phisher/server/graphql/types/ssl.py", line 97, in resolve_status
    return SSLStatus.resolve(info, **kwargs)
  File "/opt/king-phisher/king_phisher/server/graphql/types/ssl.py", line 81, in resolve
    has_letsencrypt=letsencrypt.get_certbot_bin_path(server_config),
  File "/opt/king-phisher/king_phisher/server/letsencrypt.py", line 168, in get_certbot_bin_path
    if not os.path.isfile(bin_path):
  File "/opt/king-phisher/.venv/lib/python3.6/genericpath.py", line 30, in isfile
    st = os.stat(path)
graphql.error.located_error.GraphQLLocatedError: stat: path should be string, bytes, os.PathLike or integer, not NoneType
```
Client Side:
```
ERROR    async rpc method: graphql encountered an error
Traceback (most recent call last):
  File "/opt/king-phisher/king_phisher/client/client_rpc.py", line 242, in _async_thread_routine
    results = work_item.method(*args, **kwargs)
  File "/opt/king-phisher/king_phisher/client/client_rpc.py", line 364, in graphql
    query_vars=query_vars
king_phisher.errors.KingPhisherGraphQLQueryError
```

**Comment by zeroSteiner** [comment]
Alright the underlying GraphQL error that was causing the SSL status to not be populated is fixed in commit bc0d47789fdc1e40ee240e90abac6701e053bfcb.

Normally we would tell you to opt into beta testing to get the changes, but we're in the process of doing a new release so it should be out later this week.

**Comment by Allen-Paul** [comment]
thank you
